### PR TITLE
Improve AWS SQS Sensor (#16880)

### LINKED ISSUE #16880: Improve AWS SQS Sensor
**Description**

Improve the AWS SQS Sensor as follows:
+ Add optional visibility_timeout parameter
+ Add a customisable / overrideable filter capability so we can filter/ignore irrelevant messages

[Not needed, see below conversation] 
--- Check the HTTP status code in AWS response and raise Exception if not 200 - best practice

**Use case / motivation**

I'd like to make the SQS sensor more flexible to enable the following use case:
+ A single queue can be used as a channel for messages from multiple event sources and or multiple targets
+ We need a way to filter and ignore messages not relevant to us, which other processes are looking for

**Are you willing to submit a PR?**

Yes, please assign to me

**Related Issues**

None

### PR DESCRIPTION
**Pull request by baolsen**
**Improve the AWS SQS Sensor as follows**

- Add optional visibility_timeout parameter
- Add a customisable / overrideable filter capability so we can filter/ignore irrelevant messages

closes: #16880

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -115,6 +115,7 @@ install_requires =
     # Logging is broken with itsdangerous > 2
     itsdangerous>=1.1.0, <2.0
     jinja2>=2.10.1,<4
+    jsonpath_ng>=1.5.3
```
Comment (baolsen) [inline setup.cfg]: If there's another lib we prefer for JSONPath expression parsing let me know and I'll try to rework this to use it. This was the best one I could find.

Comment (potiuk) [reply setup.cfg]: We already used it in the past and **just** removed it few weeks ago as it was not needed any more ;). So I think it's good to go.

Comment (potiuk) [reply setup.cfg]: However, rather than in setup.cfg it should be added in setup.py for amazon provider.https://github.com/apache/airflow/blob/df0746e133ca0f54adb93257c119dd550846bb89/setup.py#L182

**Review by potiuk — CHANGES_REQUESTED** [review]
LGTM, just one small comment to move the jsonpath_ng dependency to `amazon` specific provider dependencies.

File: airflow/providers/amazon/aws/sensors/sqs.py [hunk 2]
```diff
@@ -69,31 +89,48 @@ def poke(self, context):
...
+
+        response = sqs_conn.receive_message(**receive_message_kwargs)
+
+        if "Messages" not in response:
+            return False
 
-        self.log.info("received message %s", str(messages))
+        messages = response['Messages']
+        num_messages = len(messages)
+        self.log.info("received %s messages", str(num_messages))
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/sqs.py]:
```suggestion
        self.log.info("received %d messages", num_messages)
```

The string conversion is redundant since the logger can accept int directly; passing the int also ensures the most context is retained, which can be useful for people using custom log handlers and formatters. (I know this was done in existing code; it is wrong.)

Same for other logs below.

File: airflow/providers/amazon/aws/sensors/sqs.py [hunk 3]
```diff
@@ -69,31 +89,48 @@ def poke(self, context):
...
 
-        if 'Messages' in messages and messages['Messages']:
-            entries = [
-                {'Id': message['MessageId'], 'ReceiptHandle': message['ReceiptHandle']}
-                for message in messages['Messages']
-            ]
+        if num_messages == 0:
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/sqs.py]:
```suggestion
        if not num_messages:
```

Same for other checks below.

File: airflow/providers/amazon/aws/sensors/sqs.py [hunk 4]
```diff
@@ -69,31 +89,48 @@ def poke(self, context):
...
+            return False
 
-            result = sqs_conn.delete_message_batch(QueueUrl=self.sqs_queue, Entries=entries)
+        if self.message_filtering:
+            messages = self.filter_messages(messages)
+            num_messages = len(messages)
+            self.log.info("filtered %s messages", str(num_messages))
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/sqs.py]: This message is quite ambiguous. Is the count indicating how many message are filtered (i.e. dropped), or how many are left after the filter is applied?

**Review by uranusjr — CHANGES_REQUESTED** [review]

File: setup.cfg (hunk 1 above)
Comment (baolsen) [reply setup.cfg]: Fixed up, thank you for the feedback

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 2 above)
Comment (baolsen) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Thanks for the feedback, I didn't know about this. I'll get them all fixed up

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 4 above)
Comment (baolsen) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Good catch, thank you. I will make it clearer.

**Comment by baolsen** [comment]
Review comments have been addressed and I rebased with main.

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 2 above)
Comment (baolsen) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Done

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 3 above)
Comment (baolsen) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Done, thank you

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 4 above)
Comment (baolsen) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Done

File: airflow/providers/amazon/aws/sensors/sqs.py [hunk 5]
```diff
@@ -48,13 +68,26 @@ def __init__(
         aws_conn_id: str = 'aws_default',
         max_messages: int = 5,
         wait_time_seconds: int = 1,
+        visibility_timeout: Optional[int] = None,
+        message_filtering: Optional[str] = None,
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/sqs.py]:
```suggestion
        message_filtering: Optional[Literal["literal", "jsonpath"]] = None,
```

File: airflow/providers/amazon/aws/sensors/sqs.py [hunk 6]
```diff
@@ -48,13 +68,26 @@ def __init__(
...
+        message_filtering_match_values: Optional[Any] = None,
+        message_filtering_config: Optional[Any] = None,
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/sqs.py]: `Optional[Any]` is just `Any`, so the Optional part is useless

File: airflow/providers/amazon/aws/sensors/sqs.py [hunk 7]
```diff
@@ -102,3 +152,39 @@ def get_hook(self) -> SQSHook:
...
+            return self.filter_messages_literal(messages)
+        if self.message_filtering == 'jsonpath':
+            return self.filter_messages_jsonpath(messages)
+        else:
+            raise NotImplementedError('Override this method to define custom filters')
+
+    def filter_messages_literal(self, messages):
+        filtered_messages = []
+        if self.message_filtering_match_values is None:
+            raise Exception('message_filtering_match_values must be specified for literal matching')
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/sqs.py]: This should be checked in `__init__`, and shoul raise a `TypeError` instead.

File: airflow/providers/amazon/aws/sensors/sqs.py [hunk 8]
```diff
@@ -48,13 +68,26 @@ def __init__(
...
         self.aws_conn_id = aws_conn_id
         self.max_messages = max_messages
         self.wait_time_seconds = wait_time_seconds
+        self.visibility_timeout = visibility_timeout
+
+        self.message_filtering = message_filtering
+        if message_filtering_match_values is not None:
+            if not isinstance(message_filtering_match_values, list):
+                message_filtering_match_values = [message_filtering_match_values]
+        self.message_filtering_match_values = message_filtering_match_values
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/sensors/sqs.py:90]: I wonder `message_filtering_match_values` should be a `set` instead, this can improve performance quite a bit since the check is in a pretty tight loop.

**Review by uranusjr — CHANGES_REQUESTED** [review]

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 8 above)
Comment (baolsen) [reply airflow/providers/amazon/aws/sensors/sqs.py:90]: Thanks for the feedback, I did not think of that. Fixed

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 7 above)
Comment (baolsen) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Fixed, thanks for the feedback

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 6 above)
Comment (baolsen) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Noted, thank you. Fixed

File: airflow/providers/amazon/aws/sensors/sqs.py (hunk 5 above)
Comment (baolsen) [reply airflow/providers/amazon/aws/sensors/sqs.py]: Fixed, thank you

**Comment by baolsen** [comment]
Review comments 2 have been addressed and I rebased with main.

**Comment by baolsen** [comment]
Random CI issues. Rebasing & pushing again

**Review by uranusjr — APPROVED** [review]
Thanks for the fixups! (I don’t know enough about SQS to comment further)

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
There is a real error: `  ImportError: cannot import name 'Literal'`

**Review by potiuk — CHANGES_REQUESTED** [review]
Literal cannot be imported.

**Comment by baolsen** [comment]
Thanks @potiuk . Must have missed that, will get it fixed. 
I've been checking the CI but it seems quite flaky. Makes it hard to spot the signal from the noise. Is it just me or a more general issue?

**Comment by potiuk** [comment]
> Thanks @potiuk . Must have missed that, will get it fixed.
> I've been checking the CI but it seems quite flaky. Makes it hard to spot the signal from the noise. Is it just me or a more general issue?

Recently we had more issues than usual and working on getting the noise out .. Sorry for that, it's difficult to keep such a big project with such many dependencies in "excellent" shape all the times, but recently the shape was indeed "bad". Hopefully we will move it to "pretty good", and there will be hopefully long times of "really good". This takes a lot of time not only to get there, but also to keep it there.

**Comment by potiuk** [comment]
Seems like there is mostly recurring problem with  Integration test - one more rebase could probably give an indication if is a real  error

**Comment by baolsen** [comment]
Great. I've rebased & pushed. 
Ref for the previously failed CI build for comparison purposes:
https://github.com/apache/airflow/pull/16904/checks?check_run_id=3208579465

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
LGTM. And tests seem to be intermittent.

**Comment by baolsen** [comment]
Hey @potiuk thanks for the review and merge.

I have one quick question regarding CI. 
See attached screenshot from a different PR. 

It seems that the "wait for images" job is running while the "build images" job is queued. 
I would have thought the wait for images can't run until build images is done.... What am I missing?

`ci.yml` has `ci-images` needs `build-info`, but this isn't what I'm seeing.

image

**Comment by potiuk** [comment]
The problem is that we can't do it differently until we have better support from GitHub for cross-workflow communication. 

All details are described in CI.rst but here is the gist of it :

* The PR workflows have by definition only 'read ' access to the GitHub Repo of Airflow.  They cannot have 'write' access because it opens up security issue possibility ( where anyone running a PR could modify Airflow Repo)

* On the other hand we need to build (and share) ci docker images from the incoming PR so that all jobs can use those. This is needed because we have different kinds of jobs (tests/building providers / testing images / helm tests etc. - having common CI docker images with all the sources and what is more important dependencies built in make it possible to have common 'execution' environment for all tests. The problem with it is that if we have read only access in PR workflow, we cannot build such image only once and share it with other jobs because we cannot store the image anywhere. Each build job runs on separate machine and ideally you just build the images once, push it somewhere and every job uses this image. That could save a LOT of build time - especially when someone adds s new dependency (which does not happen very often but often enough)

* Therefore we run building image in separate workflow (pull_request_target) where we have write access, but also ci workflow and scripts are coming fron 'main' branch - so no risk of someone's PR injecting something in our repo. Those 'Build image' workflows can build and push image with write permissions.

* The 'pull request targer' workflow is triggered in parallel to 'ci' workflow. currently GitHub Action does not have a feature to add any 'cross workflew' dependency so we cannot wait in one workflow for the result of another (explicitly). As a workaround 'wait for image' job from ci workflow simply actively polls for the images produced by 'build image' workflow. It is a waste. Active job running and checking if images are already pushed. But there is no other way. Action do not yet have a possibility of 'pausing' and 'resuming' workflows. It is planned but not yet there 

* Also 'build image' workflow is 'safe' so we run those using our self-hosted runners while the 'ci build' is potentially unsafe (GitHub detected people using actions to mine Bitcoin(!) )  so we need to run them using Public GitHub Runners which get isolation and monitoring from GiyHub. There we have 150 slots in queue for all Apache projects (> 350 of them) which make those jobs susceptible to long queues. On the other hand we also limit our 'build image' job queue as we have limited funds from Astronomer and AWS to run those jobs in Amazon infrastructure (GCP also donated credits but we need to switch to GCP). That's why those queues run at different capacity and one is sometimes faster than the other ( in either direction).


In short 'build' and 'wait' are designed to run in parallel in any sequence and we have no way currently to enforce the sequence they run in.

I hope it makes things clearer :) (or maybe reveal the hidden complexities we are dealing with )
